torch/csrc/jit/codegen/cuda/kernel_cache.h

```diff
@@ -18,39 +19,136 @@ namespace fuser {
+  //! FusionExecutorCache API for evicting an input id
+  void evictCache(size_t input_id) {
     for (auto& fe : executors_) {
-      fe.evictCache(cache_id);
+      fe.evictCache(input_id);
     }
   }
 
-  std::vector<at::Tensor> runWithInput(const at::ArrayRef<IValue>& inputs);
+  //! FusionExecutorCache API for running the segmented fusion with given global
+  //! inputs
+  std::vector<at::Tensor> runWithInput(
+      const at::ArrayRef<IValue>& inputs,
+      size_t input_id);
+
+  //! Cache Interface: Common utility for computing hash of scheduler entires
+  static HashType getHash(SegmentHeuristics* sh);
+
+  //! Cache Interface: trivially copied and easily compared
+  //!   descriptor for FusionSegmentRuntime
+  class EntryTag {
+   public:
+    //! Computes hash upon creation
+    explicit EntryTag(SegmentHeuristics*);
+
+    //! Tag equal abstracts the heuristics equivalence
+    bool operator==(const EntryTag& other) const;
+
+    //! Returns computed hash value
+    HashType hash() const {
+      return hash_;
+    }
+
+   private:
+    HashType hash_;
+    SegmentHeuristics* heuristics_;
+  };
+
+  class EntryTagHash {
+   public:
+    HashType operator()(const EntryTag& et) const {
+      return et.hash();
+    }
+  };
 
  private:
+  //! Run one segment of the segmented fusion, compiles if not done so
   std::vector<at::Tensor> runSegmentWithInput(
       SegmentedGroup* sg,
-      const at::ArrayRef<IValue>& inputs);
+      const at::ArrayRef<IValue>& inputs,
+      size_t input_id);
+
+  //! Accessor class for the internal schedulers maintained in this runtime
+  const std::vector<SchedulerEntryPtr>& schedulers();
 
  private:
-  // Entries indexed by groupID
+  friend class EntryTag;
+  //! Entries indexed by groupID:
+  //! Executors holding compiled kernels
   std::vector<FusionExecutor> executors_;
-  std::vector<LaunchParams> launch_params_;
-  std::vector<ReductionParams> schedule_heuristics_;
+
+  //! Heuristics object holding scheduler entries for all segments
+  std::unique_ptr<SegmentHeuristics> heuristics_;
 
   // States
-  SegmentedFusion* segmented_fusion_;
-  bool compiled_ = false;
   size_t cache_id_ = -1;
+  SegmentedFusion* segmented_fusion_;
+};
+
+//! Object holding cache entries for segmented fusion
+class TORCH_CUDA_API FusionSegmentRuntimeCache {
+ public:
+  explicit FusionSegmentRuntimeCache() = default;
+
+  //! Evict the cacheEntry by id.
+  //!  removes ID to RT lookup and corresponding
+  //!  input entries. Doesn't actually release any compiled
+  //!  kernel because compiling is expensive
+  void evictId(size_t input_id);
+
+  //! In case of cache hit by input id, return pointer to that entry,
+  //!  returns nullptr if input_id miss
+  FusionSegmentRuntime* getRTById(size_t input_id);
+
+  //! In case of input id miss, evaluate heuristics and find a hit by heuristics
+  //!   in case of heuristics miss, create a new entry
+  FusionSegmentRuntime* getRTByHeuristics(
```

Comment:
Seems like this could provide a good opportunity in the future to cache on  individual segment heuristics instead of always having to redo the segmentation.